pylint/checkers/bad_chained_comparison.py

```diff
@@ -0,0 +1,58 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+from astroid import nodes
+
+from pylint.checkers import BaseChecker
+from pylint.interfaces import HIGH
+
+if TYPE_CHECKING:
+    from pylint.lint import PyLinter
+
+COMPARISON_OP = frozenset(("<", "<=", ">", ">=", "!=", "=="))
+IDENTITY_OP = frozenset(("is", "is not"))
+MEMBERSHIP_OP = frozenset(("in", "not in"))
+
+
+class BadChainedComparisonChecker(BaseChecker):
+    """Checks for unintentional usage of chained comparison."""
+
+    name = "bad-chained-comparison"
+    msgs = {
+        "W3501": (
+            "Suspicious %s-part chained comparison using semantically incompatible operators (%s)",
+            "bad-chained-comparison",
+            "Used when there is a chained comparison where one expression is part of two comparisons that belong to different semantic groups "
+            ' ("<" does not mean the same thing as "is", chaining them in "0 < x is None" is probably a mistake).',
+        )
+    }
+
+    def _get_distinct_operators(self, node: nodes.Compare) -> list:
+        operators = [op[0] for op in node.ops]
+        return list(dict.fromkeys(operators))

```

Comment:
```suggestion
```
We're trying to get a list of uniq operator right ? Shouldn't this be a set as in ``set(op[0] for op in node.ops)`` ?

Reply:
Using `set(...)` gives us a list of unique operators, however, the operators will be unordered (from my understanding, python sets are unordered). Therefore, the message output will be non-deterministic.

The `list(dict.fromkeys(...))` gives a list of unique operators as well, and will order the operators according to their position in the expression https://stackoverflow.com/questions/39755464/python-unique-items-of-list-in-the-order-that-it-appears. 
For example: given the expression `(a < b <= c in d)`, the list will always be `['<', '<=', 'in']`. Whereas the set may create `{'<=', 'in', '<'}` in an instance, and output `{'in', '<', '<='}` in another instance.

In my opinion, it would be better if the message output is deterministic. Also, so that we can perform functional testing 🙂 Let me know what you think.

Reply:
Maybe we can sort explicitly then ?

Reply:
that works too, since the operator set is quite small, sorting shouldn't be too costly. let's go with sort so that the code will be easier to understand.